src/lib/nested_property.js

```diff
@@ -115,44 +113,21 @@ function npGet(cont, parts) {
 }
 
 /*
- * Can this value be deleted? We can delete any empty object (null, undefined, [], {})
- * EXCEPT empty data arrays, {} inside an array, or anything INSIDE an *args* array.
+ * Can this value be deleted? We can delete `undefined`, and `null` except INSIDE an
+ * *args* array.
  *
- * Info arrays can be safely deleted, but not deleting them has no ill effects other
- * than leaving a trace or layout object with some cruft in it.
+ * Previously we also deleted some `{}` and `[]`, in order to try and make set/unset
+ * a net noop; but this causes far more complication than it's worth, and still had
+ * lots of exceptions. See https://github.com/plotly/plotly.js/issues/1410
  *
- * Deleting data arrays can change the meaning of the object, as `[]` means there is
- * data for this attribute, it's just empty right now while `undefined` means the data
- * should be filled in with defaults to match other data arrays.
- *
- * `{}` inside an array means "the default object" which is clearly different from
- * popping it off the end of the array, or setting it `undefined` inside the array.
- *
- * *args* arrays get passed directly to API methods and we should respect precisely
- * what the user has put there - although if the whole *args* array is empty it's fine
- * to delete that.
- *
- * So we do some simple tests here to find known non-data arrays but don't worry too
- * much about not deleting some arrays that would actually be safe to delete.
+ * *args* arrays get passed directly to API methods and we should respect null if
+ * the user put it there, but otherwise null is deleted as we use it as code
+ * in restyle/relayout/update for "delete this value" whereas undefined means
+ * "ignore this edit"
  */
-var INFO_PATTERNS = /(^|\.)((domain|range)(\.[xy])?|args|parallels)$/;
```

Comment:
TODO:

- [ ] check that the old toolpanel does not break

Reply:
Moreover @alexcjohnson, what if in the very unlikely scenario of someone writing in that their app relies on the old pruning behavior, what should we do?

I suspect reverting this commit would break some test(s) added in this PR (which one by the way?), so maybe we could formulate a solution for that hypothetical user?

Reply:
The reason I made this commit now was the `trace._length = null`, which in some cases was getting pruned later on in `supplyDefaults` by attributes with `dflt: null` and breaking my [redraw all the things test addition](https://github.com/plotly/plotly.js/pull/2577/files#diff-5a140ce1cc87a420496039bbf8d699f1R3276). I could have handled this in a few different ways:
- use something other than `null` for `_length` - I considered `false` but to me that had the implication of "has no length" rather than what I wanted, "(1D) length does not apply." Anyway the whole idea that we're pruning in the middle of `supplyDefaults` which is really only supposed to be *building* the `_full*` objects I wanted to avoid.
- make a special code path that does no pruning or unsetting - like `nestedProperty.onlySet(val)` - but that's adding complexity so not ideal.
- get rid of all `dflt: null` (and perhaps even enforce that) throughout our attributes. I still think we should do that, but I opted not to do it here.

Anyway, to the question:
> what if in the very unlikely scenario of someone writing in that their app relies on the old pruning behavior, what should we do?

If it leads to a visible change in the plot, we should fix it so the API does not contain such a dependence on the existence of empty containers ;)
If it just alters the data structures (`data` and `layout` I guess, after a `restyle`/`relayout`) I can't really predict why a user would be bothered by this, but we can help them wrap their `restyle`/`relayout` with something like `if(val === null) { pruneContainersPulledFromThisCommit(...) }`

Reply:
Great answer. Thanks!